DjangoProject/bookmarksapp/tests.py

```diff
@@ -6,69 +6,151 @@
 
 
 class BookmarksAppTestCase(TestCase):
+    user = None
 
     def setUp(self):
+        """
+        This function will execute before for each testcase.
+        User creation and login will happen.
+        Some sample bookmarks and folders will be save in database.
+        """
         self.credentials = {
             'username': 'pavan@gmail.com',
             'password': 'pavankumar'
         }
         MyUser.objects.create_user(**self.credentials)
-        tempuser = self.client.post(reverse('authenticationapp:login'), self.credentials, follow=True).context['user']
+        self.user = self.client.post(reverse('authenticationapp:login'), self.credentials, follow=True).context['user']
         f = Folders.objects.create(
             name="Google",
-            created_by=tempuser
+            created_by=self.user
         )
         y = Folders.objects.create(
             name="Yahoo",
-            created_by=tempuser
+            created_by=self.user
         )
 
         Bookmarks.objects.create(
             folder=f,
             url="https://www.google.com",
             name='Google Website',
-            created_by=tempuser
+            created_by=self.user
         )
 
         Bookmarks.objects.create(
             folder=y,
             url="https://www.yahoo.com",
             name='Yahoo Website',
-            created_by=tempuser
+            created_by=self.user
         )
 
-    def testFolderListWithoutLogin(self):
+    def tearDown(self):
+        """
+        Logout the user after finishing every testcase.
+        """
+        self.client.get(reverse('authenticationapp:logout'))
+
+    def test_folder_list_without_login(self):
+        """
+        Test folderlist page without logged in the user.
+        """
         self.client.get(reverse('authenticationapp:logout'))
```

Comment:
Why logging out at the start of each test case? It should be taken care of by the teardown.

Reply:
For this testcase, I need to check which page it is rendering if user is not logged in. Since user is already logged in with setUp method and teardown will execute after completing this testcase. I am logging out inside this testcase.